teps/0125-configmaps-and-secrets-as-param-value-source.md

```diff
@@ -0,0 +1,189 @@
+
+---
+title: Configmap and Secret as Param Value Source
+authors:
+  - "@rpajay"
+ 
+creation-date: 2022-10-28\
+last-updated: 2022-10-28\
+status: pending
+
+---
+# TEP-0110: Configmap and Secret as Taskrun or Pipelinerun Param's Value Source
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Contract](#contract)
+  - [API](#api)
+  - [User Stories (optional)](#user-stories-optional)
+    - [Versioned <code>Task</code>s and <code>Pipeline</code>s and Pipeline-as-code](#versioned-s-and-s-and-pipeline-as-code)
+    - [Shipping catalog resources as OCI images](#shipping-catalog-resources-as-oci-images)
+    - [Tooling](#tooling)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+This proposal is to be able to reference `ConfigMap` or `Secret` as value source for `TaskRun` or `PipelineRun` `Params`.\
+This is to support Kubernetes native options (ConfigMap or Secret) as value source along with direct value passed to `TaskRun` and `PipelineRun`.
+
+## Motivation
+
+`PipelineRun` and `TaskRun` Params support passing direct value during its creation. Two supported options to pass values from Secret / ConfigMap are
+- `TEP 0029-step-workspaces`  where Secret / ConfigMap is mounted as file inside container.
+- Assigning `Env` from `ConfigMap / Secret` reference for each container / step on creation of pipeline.
+
+To understand the problem let's take an example. 
+jib-maven[https://github.com/tektoncd/catalog/blob/main/task/jib-maven/0.4/jib-maven.yaml]
+
+```yaml
+steps:
+	- name: build-and-push
+	  image: $(params.MAVEN_IMAGE)
+	  # Make sh evaluate $HOME.
+	  script: |
+		#!/bin/bash
+		[[ -f /tekton-custom-certs/$(params.CACERTFILE) ]] && \
+		keytool -import -keystore $JAVA_HOME/lib/security/cacerts -storepass "changeit" -file /tekton-custom-certs/$(params.CACERTFILE) -noprompt
+		mvn -B \
+		-Duser.home=$HOME \
+		-Djib.allowInsecureRegistries=$(params.INSECUREREGISTRY) \
+		-Djib.to.image=$(params.IMAGE) \
+		compile \
+		com.google.cloud.tools:jib-maven-plugin:build
+```
+
+In this tekton hub task where 2 params are
+- `MAVEN_IMAGE` : Either of the above mentioned solution  doesn't work as the value isn't a param but as an env variable inside the container
+- `CACERTFILE` : Secret doesn't work out of the box as to access env variable `$(params.CACERTFILE)` needs to be replaced with `$CACERTFILE`
```

Comment:
The `Task` in question is expanding params directly in a script, which is a bad practice because it exposes scripts to malicious injections.

A better implementation would be:

```yaml
        steps:
        - name: build-and-push
          image: $(params.MAVEN_IMAGE)
          env:
          - name: CACERTFILE
            value: $(params.CACERTFILE)
          - name: INSECUREREGISTRY
            value: $(params.INSECUREREGISTRY)
          - name: IMAGE
            value: $(params.IMAGE)
          # Make sh evaluate $HOME.
          script: |
                #!/bin/bash
                [[ -f "/tekton-custom-certs/${CACERTFILE}" ]] && \
                keytool -import -keystore $JAVA_HOME/lib/security/cacerts -storepass "changeit" -file "/tekton-custom-certs/${CACERTFILE}" -noprompt
                mvn -B \
                -Duser.home=$HOME \
                -Djib.allowInsecureRegistries=${INSECUREREGISTRY} \
                -Djib.to.image=${IMAGE} \
                compile \
                com.google.cloud.tools:jib-maven-plugin:build
```

The `env` in the example above could actually load values from a config map or secret, but that ability would have to be implemented in the `Task` and it would replace the ability to pass value directly via parameters, which is not very portable and user friendly.

Reply:
This is just an example from tekton hub :)